misc/incremental_checker.py

```diff
@@ -0,0 +1,218 @@
+        delta, output = run_mypy(target)
+        expected_delta = cache[commit_id]['delta']  # type: float
+        expected_output = cache[commit_id]['output']  # type: str
+        if output != expected_output:
+            print("    Output does not match expected result!")
+            print("    Expected output ({:.3f} sec):".format(expected_delta))
+            print_offset(expected_output, 8)
+            print("    Actual output: ({:.3f} sec):".format(delta))
+            print_offset(output, 8)
+        else:
+            print("    Output matches expected result!")
+            print("    Incremental: {:.3f} sec".format(delta))
+            print("    Original:    {:.3f} sec".format(expected_delta))
+
+
+def cleanup(folder_path: str) -> None:
+    print("Cleanup: deleting {}".format(folder_path))
+    delete_folder(folder_path)
+    delete_folder("misc/.mypy_cache")
+
+
+def test_repo(target_repo_url: str, module_name: str, start_commit: str,
+              branch: str, cache_path: str) -> None:
+    repo_folder_name = "tmp_repo"
+    target = os.path.join(repo_folder_name, module_name)
+
+    # Assume we're starting in misc
+    os.chdir("..")
+    initialize_repo(target_repo_url, repo_folder_name, branch)
+
+    if start_commit is None:
+        start_commit = get_nth_commit(repo_folder_name, LOOK_BACK)[0]
+    commits = get_commits_starting_at(repo_folder_name, start_commit)
+
+    cache = load_cache(cache_path)
+    set_expected(commits, cache, repo_folder_name, target)
+    save_cache(cache, cache_path)
+
+    test_incremental(commits, cache, repo_folder_name, target)
+
+    cleanup(repo_folder_name)
+
+
+def main() -> None:
+    help_factory = (lambda prog: RawDescriptionHelpFormatter(prog=prog, max_help_position=32))
+    parser = ArgumentParser(
+        prog='incremental_checker',
+        description=__doc__,
+        formatter_class=help_factory)
+
+    parser.add_argument("-r", "--repo_url", default=MYPY_REPO_URL, metavar="URL",
+                        help="the repo to clone and run tests on")
+    parser.add_argument("-m", "--module-name", default=MYPY_MODULE_NAME, metavar="NAME",
+                        help="the name of the module to typecheck")
+    parser.add_argument("-C", "--commit-id", default=None, metavar="ID",
+                        help="the commit id to start from "
+                        "(starts {} commits back if missing)".format(LOOK_BACK))
+    parser.add_argument("--cache-path", default=CACHE_PATH, metavar="DIR",
+                        help="sets a custom location to store cache data")
+    parser.add_argument("--branch", default=None, metavar="NAME",
+                        help="check out and test a custom branch"
+                        "uses the default if not specified")
+
+    params = parser.parse_args(sys.argv[1:])
+    cache_path = os.path.abspath(params.cache_path)
+    test_repo(params.repo_url, params.module_name, params.commit_id, params.branch, cache_path)
```

Comment:
I ran the script without args and was a little surprised to find that it immediately started chugging along. I wonder if without any arguments it should just spit out the help text? You'd have to specify at least one argument to get it to run.